app/rust/src/lib.rs

```diff
@@ -0,0 +1,232 @@
+*  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+*  See the License for the specific language governing permissions and
+*  limitations under the License.
+********************************************************************************/
+#![no_std]
+#![no_main]
+#![no_builtins]
+#![allow(dead_code, unused_imports)]
+
+use core::panic::PanicInfo;
+
+use constants::{DIV_DEFAULT_LIST_LEN, DIV_SIZE, SPENDING_KEY_GENERATOR, KEY_DIVERSIFICATION_PERSONALIZATION, GH_FIRST_BLOCK};
+mod constants;
+use aes::Aes256;
+use aes::cipher::{
+    BlockCipher, BlockEncrypt, BlockDecrypt, NewBlockCipher,
+    generic_array::{GenericArray,typenum::U32},
+};
+use binary_ff1::BinaryFF1;
+use jubjub::{AffinePoint, ExtendedPoint, Fr};
+use blake2s_simd::{blake2s, Hash as Blake2sHash, Params as Blake2sParams};
+
+fn debug(_msg: &str) {}
+
+// ParserError should mirror parser_error_t from parser_common.
+// At the moment, just implement OK or Error
+#[repr(C)]
+pub enum ParserError {
+    ParserOk = 0,
+    ParserUnexpectedError = 5,
+}
+
+#[repr(C)]
+pub enum ConstantKey {
+    SpendingKeyGenerator,
+    ProofGenerationKeyGenerator,
+}
+
+#[no_mangle]
+pub extern "C" fn from_bytes_wide(input: &[u8; 64], output: &mut [u8; 32]) -> ParserError {
+    let result = Fr::from_bytes_wide(input).to_bytes();
+    output.copy_from_slice(&result[0..32]);
+    ParserError::ParserOk
+}
+
+#[no_mangle]
+pub extern "C" fn scalar_multiplication(
+    input: &[u8; 32],
+    key: ConstantKey,
+    output: *mut [u8; 32],
+) -> ParserError {
+    let key_point = match key {
+        ConstantKey::SpendingKeyGenerator => constants::SPENDING_KEY_GENERATOR,
+        ConstantKey::ProofGenerationKeyGenerator => constants::PROOF_GENERATION_KEY_GENERATOR,
+    };
+
+    let extended_point = key_point.multiply_bits(input);
+    let result = AffinePoint::from(&extended_point);
+
+    unsafe {
+        let output_slice = &mut *output;
+        output_slice.copy_from_slice(&result.to_bytes());
+    }
+
+    ParserError::ParserOk
+}
+
+#[inline(never)]
+pub fn get_diversifiers(
+    dk: &[u8; 32],
+    start_index: &mut [u8; 11],
+    result: &mut [u8; 44],
+) {
+    // Initialize cipher
+    let key = GenericArray::from_slice(dk);
+    let cipher = Aes256::new(&key);
+    let mut scratch = [0; 12];
+
+    let mut ff1 = BinaryFF1::new(&cipher, 11, &[], &mut scratch).unwrap();
+
+    let mut d: [u8; 11];
+    let size = 4;
+
+    for c in 0..size {
+        d = *start_index;
+        ff1.encrypt(&mut d).unwrap();
+        result[c * 11..(c + 1) * 11].copy_from_slice(&d);
+        for k in 0..11 {
+            start_index[k] = start_index[k].wrapping_add(1);
+            if start_index[k] != 0 {
+                // No overflow
+                break;
+            }
+        }
+    }
+}
+
+#[inline(never)]
+fn diversifier_group_hash_light(tag: &[u8]) -> bool {
+    if tag == [0u8; 11] {
+        return false;
+    }
+    let h = Blake2sParams::new()
```

Comment:
this function can be moved to C. I think that we only need to run lines 122~16 in Rust, everything else can be done in C and remove Blake2s hash from Rust deps (if it's not used in any other place)